aten/src/ATen/TensorSubclassLikeUtils.h

```diff
@@ -0,0 +1,48 @@
+#pragma once
+#include <ATen/ATen.h>
+
+namespace at {
+
+// Note [Tensor-subclass-like Tensors]
+// Tensor-subclass-like is defined as:
+// - a Tensor subclass (via __torch_dispatch__ in Python or extending
+//   TensorImpl in C++)
+// - anything else that shares the same perils as Tensor subclasses.
+//   For example, many Tensor subclasses do not have storage and meta Tensors
```

Comment:
This should also include sparse no? And we can use this to replace sparse checks in a bunch of places (thinking about the autograd engine)

Reply:
Should it include sparse? Sparse tensors have additional constraints: in addition to (1) no data_ptr and (2) no incompatible in-place, sparse tensors also have restrictions around what view operations may be applied to them.

Reply:
So they have even more restrictions? they should at least be here right?

Reply:
Yes, they have even more restrictions :(. We can certainly add them here and add extra sparse checks

Reply:
I added four of the sparse dispatch keys. There are three that I don't know if I should have added:

https://github.com/pytorch/pytorch/blob/baac51ff4ad492c3c89d4b189a57543a89ae2b3d/c10/core/DispatchKey.h#L111-L114

SparseHIP seems to be dead (free dispatch key?) and SparseXPU/SparseVE are out of tree so I'd want their developers to come to us and say they want this feature before opting them into it.

At any rate we can always add more to the list